### PR TITLE
website: Expand data handling documentation with type-specific pages and examples

### LINKED ISSUE #695: Make value handling methods more discoverable in the documentation
### Module version

N/A

### Use-cases

As a new framework developer, I would like to easily find the methods for converting go native types to their equivalent `type` values.

The current [documenation](https://developer.hashicorp.com/terraform/plugin/framework/handling-data/attributes) for attributes is very long and the full contents are not listed completely on the "On this page".

### Attempted Solutions

N/A

### Proposal

1. Add a general "`types` Conventions" section at the top of the page describing that each type has a "types.<type>.IsNull"/"types.<type>ValueFrom"/etc.
1. Add links to the [table](https://developer.hashicorp.com/terraform/plugin/framework/handling-data/attributes#framework-attribute-types) at the top of the page to the content further down.
1. Expand the TOC (or do some other rearrangement of the content) such that each `type` has an entry in a TOC.

### PR DESCRIPTION
**Pull request by bflad**
Closes #695

This change creates new attributes, blocks, and types sections, each with full information about each type. The attribute/block pages are cross-linked with their associated type pages and vice-versa.

Web platform changes to redirect the removed pages to existing pages will be handled separately on deployment.

### CODE REVIEW

File: website/docs/plugin/framework/handling-data/attributes/number.mdx [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+
+Number attributes store an arbitrary precision (generally over 64-bit, up to 512-bit) number. Values are represented by a [number type](/terraform/plugin/framework/handling-data/types/number) in the framework.
+
+In this Terraform configuration example, an number attribute named `example_attribute` is set to the value `1.23`:
+
+```terraform
+resource "examplecloud_thing" "example" {
+  example_attribute = 1.23
+}
+```
```
Comment (bendbennett) [inline website/docs/plugin/framework/handling-data/attributes/number.mdx:23]: Nit: Do we want to consider using an example where the required precision is > 64 bit to emphasise the use-case?

File: website/docs/plugin/framework/handling-data/types/bool.mdx [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+    // ...
+}
+
+// optional logic for handling unknown value
+if data.ExampleAttribute.IsUnknown() {
+    // ...
+}
+
+// myString now contains a Go bool with the known value
+myString := data.ExampleAttribute.ValueBool()
```
Comment (bendbennett) [inline website/docs/plugin/framework/handling-data/types/bool.mdx:64]: Nit: Could use `myBool` to emphasise the type held in the var?

File: website/docs/plugin/framework/handling-data/types/map.mdx [hunk 3]
```diff
@@ -0,0 +1,127 @@
...
+* [`types.MapValueMust(map[string]attr.Type, map[string]attr.Value) types.Map`](https://pkg.go.dev/github.com/hashicorp/terraform-plugin-framework/types#MapValueMust): A known value with the given element type and values. Any diagnostics are converted to a runtime panic. This is recommended only for testing or exhaustively tested logic.
+
+In this example, a known map value is created from framework types:
+
+```go
+elements := map[string]attr.Value{
+    "key1": types.StringValue("value1"),
+    "key2": types.StringValue("value2"),
+}
+listValue, diags := types.MapValue(types.StringType, elements)
```
Comment (bendbennett) [inline website/docs/plugin/framework/handling-data/types/map.mdx:92]: Nit: Could use `mapValue` to emphasise type in var.

File: website/docs/plugin/framework/handling-data/types/map.mdx [hunk 4]
```diff
@@ -0,0 +1,127 @@
...
+```
+
+In this example, a `types.List` of `types.String` is created from a `[]string`:
+
+```go
+elements := map[string]string{
+    "key1": "value1",
+    "key2": "value2",
+}
+listValue, diags := types.MapValueFrom(ctx, types.StringType, elements)
```
Comment (bendbennett) [inline website/docs/plugin/framework/handling-data/types/map.mdx:122]: Nit: Could use `mapValue` to emphasise type in var.

**Review by bendbennett — APPROVED** [review]
Superb! This is fantastic 🚀

File: website/docs/plugin/framework/handling-data/attributes/number.mdx (hunk 1 above)
Comment (bflad) [reply website/docs/plugin/framework/handling-data/attributes/number.mdx:23]: Great suggestion -- let me find a fun one that Terraform likes.

File: website/docs/plugin/framework/handling-data/types/bool.mdx (hunk 2 above)
Comment (bflad) [reply website/docs/plugin/framework/handling-data/types/bool.mdx:64]: Good copypasta 🍝  catch, will update!

File: website/docs/plugin/framework/handling-data/types/map.mdx (hunk 3 above)
Comment (bflad) [reply website/docs/plugin/framework/handling-data/types/map.mdx:92]: Good copypasta 🍝  catch, will update!

File: website/docs/plugin/framework/handling-data/types/map.mdx (hunk 4 above)
Comment (bflad) [reply website/docs/plugin/framework/handling-data/types/map.mdx:122]: Good copypasta 🍝  catch, will update!

File: website/docs/plugin/framework/handling-data/types/map.mdx [hunk 5]
```diff
@@ -0,0 +1,127 @@
...
+
+* [`(types.Map).IsNull() bool`](https://pkg.go.dev/github.com/hashicorp/terraform-plugin-framework/types/basetypes#MapValue.IsNull): Returns `true` if the map is null.
+* [`(types.Map).IsUnknown() bool`](https://pkg.go.dev/github.com/hashicorp/terraform-plugin-framework/types/basetypes#MapValue.IsUnknown): Returns `true` if the map is unknown. Returns `false` if the number of elements is known, any of which may be unknown.
+* [`(types.Map).Elements() map[string]attr.Value`](https://pkg.go.dev/github.com/hashicorp/terraform-plugin-framework/types/basetypes#MapValue.Elements): Returns the known `map[string]attr.Value` value, or `nil` if null or unknown.
+* [`(types.Map).ElementsAs(context.Context, any, bool) diag.Diagnostics`](https://pkg.go.dev/github.com/hashicorp/terraform-plugin-framework/types/basetypes#MapValue.ElementsAs): Converts the known values into the given Go type, if possible. It is recommended to use a map of framework types to account for elements which may be unknown.
+
+In this example, a map of strings value is checked for being null or unknown value first, before accessing its known value elements as a `[]types.String`:
```
Comment (austinvalle) [inline website/docs/plugin/framework/handling-data/types/map.mdx:50]:
```suggestion
In this example, a map of strings value is checked for being null or unknown value first, before accessing its known value elements as a `map[string]types.String`:
```

File: website/docs/plugin/framework/handling-data/types/map.mdx [hunk 6]
```diff
@@ -0,0 +1,127 @@
...
+
+```go
+elements := map[string]string{
+    "key1": "value1",
+    "key2": "value2",
+}
+diags := resp.State.SetAttribute(ctx, path.Root("example_attribute"), elements)
+```
+
+In this example, a `types.List` of `types.String` is created from a `[]string`:
```
Comment (austinvalle) [inline website/docs/plugin/framework/handling-data/types/map.mdx:115]:
```suggestion
In this example, a `types.Map` of `types.String` is created from a `map[string]string`:
```

File: website/docs/plugin/framework/handling-data/types/object.mdx [hunk 7]
```diff
@@ -0,0 +1,170 @@
+---
+page_title: 'Plugin Development - Framework: Object Type'
+description: >-
+  Learn the object value type in the provider development framework.
+---
+
+# Object Type
+
+Object types store a mapping of explicit attribute names to value types. Objects must declare all attribute values, even when null or unkown, unless the entire object is null or unknown.
```
Comment (austinvalle) [inline website/docs/plugin/framework/handling-data/types/object.mdx:9]:
```suggestion
Object types store a mapping of explicit attribute names to value types. Objects must declare all attribute values, even when null or unknown, unless the entire object is null or unknown.
```

**Review by austinvalle — APPROVED** [review]
Couple comments, but overall this documentation was fantastic, really like the layout! 👍🏻

File: website/docs/plugin/framework/handling-data/terraform-concepts.mdx [hunk 8]
```diff
@@ -6,48 +6,118 @@ description: >-
...
+`foo_id`.
+
+Because they can result from interpolations in the practitioner's config,
+you have no control over what attributes may contain an unknown
+value. However, by the time a resource is expected to be created, read, updated, or
+deleted, only its computed attributes can be unknown. The rest are
+guaranteed to have known values (or be null).
+
+Provider configuration values can be unknown, and providers should handle that
+situation, even if that means just returning an error.
```
[the remark this answers is not in the document]
Comment (bflad) [reply website/docs/plugin/framework/handling-data/terraform-concepts.mdx:123]: This is a very unfortunate practitioner-desired pattern, but current Terraform design anti-pattern: https://github.com/hashicorp/terraform/issues/2430 -- there's work being done in that space that cannot be discussed publicly, but there's also the public https://github.com/hashicorp/terraform/issues/30937 which noodles on whether Terraform could support wholly unknown resource data during planning, for example.

Here is a quick, pseudo-example:

```
resource "examplecloud_kubernetes_cluster" "example" {}

provider "kubernetes" {
  host = examplecloud_kubernetes_cluster.example.url
}
```

If the managed resource has not been created, its URL value is likely not known until after apply. Provider configuration does not "participate" in the operation graph like other nodes, so Terraform will call `ProviderConfigure` while passing across the unknown value. "Downstream" providers like these generally have to document workarounds:  https://registry.terraform.io/providers/hashicorp/kubernetes/latest/docs#stacking-with-managed-kubernetes-cluster-resources

If adding an example there makes it clearer, happy to add that to the moved prose there.

File: website/docs/plugin/framework/handling-data/attributes/bool.mdx [hunk 9]
```diff
@@ -0,0 +1,120 @@
...
+}
+```
+
+If the bool value should be the element type of a [collection attribute type](/terraform/plugin/framework/handling-data/attributes#collection-attribute-types), set the `ElementType` field according to the [bool type](/terraform/plugin/framework/handling-data/types/bool). Refer to the collection attribute type documentation for additional details.
+
+If the bool value should be a value type of an [object attribute type](/terraform/plugin/framework/handling-data/attributes#object-attribute-type), set the `AttributeTypes` map value according to the [bool type](/terraform/plugin/framework/handling-data/types/bool). Refer to the object attribute type documentation for additional details.
+
+### Configurability
+
+At least one of the `Computed`, `Optional`, or `Required` fields must be defined. This defines how Terraform and the framework should expect data to set, whether the value is from the pracititioner configuration or from the provider logic, such as API response value.
```
[the remark this answers is not in the document]
Comment (bflad) [reply website/docs/plugin/framework/handling-data/attributes/bool.mdx:51]: Sure thing -- will update this everywhere. 👍

File: website/docs/plugin/framework/handling-data/terraform-concepts.mdx (hunk 8 above)
[the remark this answers is not in the document]
Comment (austinvalle) [reply website/docs/plugin/framework/handling-data/terraform-concepts.mdx:123]: Ah okay, thanks for that explanation. That helps me understand, but I'm not convinced an example would really benefit provider developers. I'm good with leaving this as-is

File: website/docs/plugin/framework/handling-data/attributes/bool.mdx [hunk 10]
```diff
@@ -0,0 +1,120 @@
...
+  Learn the bool attribute type in the provider development framework.
+---
+
+# Bool Attribute
+
+Bool attributes store a boolean true or false value. Values are represented by a [bool type](/terraform/plugin/framework/handling-data/types/bool) in the framework.
+
+In this Terraform configuration example, a bool attribute named `example_attribute` is set to the value `true`:
+
+```terraform
```
[the remark this answers is not in the document]
Comment (bflad) [reply website/docs/plugin/framework/handling-data/attributes/bool.mdx:13]: For now, will convert these to `hcl`, but talking with the developer website folks about either adding a "content check"  to suggest `hcl` instead or implementing `terraform` support.

**Comment by mschuchard** [comment]
Past me from six months ago would have very much appreciated this, and therefore so will probably many others in the present and future.

**Pull request merged**